### PR TITLE
Implement TCP server

### PR DESCRIPTION
**Pull request by panferovi**

### CODE REVIEW

File: include/server/client.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+class Client {
+public:
+    void Connect();
+    void SendRequest(Request request);
+
+private:
+    boost::asio::io_service io_service_;
+    Session session_ {io_service_};
+};
```
Comment (AndreyVV-100) [inline include/server/client.h:18]: Maybe swap private and public parts and remove `private`?

Comment (AndreyVV-100) [reply include/server/client.h:18]: Same in other files.

File: include/server/session.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+class Session {
+public:
+    static constexpr uint16_t PORT = 1234;
+
+    Session(boost::asio::io_service &io_service) : socket_(io_service) {}
+
+    static Session *Create(boost::asio::io_service &io_service)
+    {
+        Session *session = new Session(io_service);
+        return session;
```
Comment (AndreyVV-100) [inline include/server/session.h:17]: return new Session(io_service);

File: include/threadpool/thread_pool.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    ~ThreadPool();
+
+    NO_COPY_SEMANTIC(ThreadPool);
+    NO_MOVE_SEMANTIC(ThreadPool);
+
+    // Launches worker threads
+    void Start();
+
+    // Schedules task for execution in one of the worker threads
+    void Submit(Task);
```
Comment (AndreyVV-100) [inline include/threadpool/thread_pool.h:28]: Variable name?

File: include/threadpool/wait_group.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
+#ifndef THREADPOOL_WAIT_GROUP_H
+#define THREADPOOL_WAIT_GROUP_H
+
+#include <mutex>
+#include <cstdlib>
```
Comment (AndreyVV-100) [inline include/threadpool/wait_group.h:5]: Is this include necessary?

File: include/utils/macros.h [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+#define LIKELY(exp) (__builtin_expect((exp) != 0, true))
+#define UNLIKELY(exp) (__builtin_expect((exp) != 0, false))
+
+#if !defined(NDEBUG)
+
+#define ALWAYS_INLINE
+
+#define LOG_DEBUG(format, ...) printf(format, ##__VA_ARGS__)
+
+#define ASSERT_FAIL(expr) shamir::AssertionFail(expr, __FILE__, __LINE__, __FUNCTION__)
```
Comment (AndreyVV-100) [inline include/utils/macros.h:51]: Fail((expr), ...)
Rework defines in all file.

File: libs/server/client.cpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#include "server/client.h"
+#include "utils/macros.h"
+
+namespace shamir {
+
+void Client::Connect()
+{
+    auto ip = boost::asio::ip::address::from_string("127.0.0.1");
```
Comment (AndreyVV-100) [inline libs/server/client.cpp:8]: Address as parameter?

File: libs/server/tcp_server.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        StartSession();
+        io_service_.run();
+    });
+}
+
+void TCPServer::Stop()
+{
+    ASSERT(IsClosed());
+    thread_pool_.Stop();
+    Session::Destroy(last_session_);
```
Comment (AndreyVV-100) [inline libs/server/tcp_server.cpp:22]: Why it's better then `delete last_session_;`?

File: libs/server/tcp_server.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+void TCPServer::Close()
+{
+    is_closed_ = true;
+    io_service_.stop();
+}
+
+bool TCPServer::IsClosed()
```
Comment (AndreyVV-100) [inline libs/server/tcp_server.cpp:31]: `inline bool TCPServer::IsClosed() const`?
Are there `const` problem in other places?

File: libs/threadpool/thread_pool.cpp [hunk 9]
```diff
@@ -0,0 +1,68 @@
+#include "threadpool/thread_pool.h"
+
+#include "utils/macros.h"
+
+namespace shamir {
+
+thread_local ThreadPool *current_pool;
+
+ThreadPool::ThreadPool(size_t threads) : threads_cnt_(threads) {}
```
Comment (AndreyVV-100) [inline libs/threadpool/thread_pool.cpp:9]: It's so strange when you write some part of code in header, but other part in cpp file.

File: libs/server/tcp_server.cpp (hunk 8 above)
Comment (AndreyVV-100) [reply libs/server/tcp_server.cpp:31]: Why we can't write this method in class?

File: libs/threadpool/thread_pool.cpp [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+{
+    while (!workers_done_.load()) {
+        auto task = task_queue_.Take();
+        if (task != std::nullopt) {
+            current_pool = this;
+            try {
+                (*task)();
+                uncompleted_tasks_.Done();
+            } catch (...) {
+                std::abort();
```
Comment (AndreyVV-100) [inline libs/threadpool/thread_pool.cpp:61]: :-)

File: include/server/client.h (hunk 1 above)
Comment (panferovi) [reply include/server/client.h:18]: no.

File: include/threadpool/queue.h [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        queue_.push_back(std::move(element));
+        is_empty_or_closed_.notify_one();
+        return true;
+    }
+
+    std::optional<T> Take()
+    {
+        std::unique_lock lock(mutex_);
+        is_empty_or_closed_.wait(lock, [this] { return !queue_.empty() || is_closed_; });
+        if (is_closed_ && queue_.empty()) {
```
[the remark this answers is not in the document]
Comment (panferovi) [reply include/threadpool/queue.h:30]: questions?

File: include/threadpool/thread_pool.h (hunk 3 above)
Comment (panferovi) [reply include/threadpool/thread_pool.h:28]: Task task

File: include/threadpool/wait_group.h (hunk 4 above)
Comment (panferovi) [reply include/threadpool/wait_group.h:5]: yes.

File: include/utils/macros.h (hunk 5 above)
Comment (panferovi) [reply include/utils/macros.h:51]: no.

File: libs/server/client.cpp (hunk 6 above)
Comment (panferovi) [reply libs/server/client.cpp:8]: Thank you for your submission. I will fix in some near future (never).

File: libs/server/tcp_server.cpp (hunk 7 above)
Comment (panferovi) [reply libs/server/tcp_server.cpp:22]: Yes! Fixed

File: libs/server/tcp_server.cpp (hunk 8 above)
Comment (panferovi) [reply libs/server/tcp_server.cpp:31]: FIxed

File: libs/threadpool/thread_pool.cpp (hunk 10 above)
Comment (panferovi) [reply libs/threadpool/thread_pool.cpp:61]: ...)

File: libs/threadpool/thread_pool.cpp (hunk 9 above)
Comment (panferovi) [reply libs/threadpool/thread_pool.cpp:9]: are u dolbaeb

**Comment by AndreyVV-100** [comment]
LGTM

**Review by AndreyVV-100 — APPROVED** [review]

**Pull request merged**